src/js/actions/documents.js

```diff
@@ -935,35 +934,70 @@ define(function (require, exports) {
-        // we reset the layer to update its smart object type and position.
-        if (layer) {
-            return this.transfer(historyActions.newHistoryStateRogueSafe, document.id)
-                .bind(this)
-                .then(function () {
-                    return this.transfer(layerActions.resetLayers, document, layer);
-                })
-                .then(function () {
-                    return this.transfer(layerActions.resetIndex, document);
+        // We only receive one "placeEvent" when dropping multiple files onto the canvas.
+        // So we need to fetch all layer IDs of the current document in order to know which layers are new.
+        return this.transfer(layerActions.getLayerIDsForDocumentID, document.id)
+            .bind(this)
+            .then(function (result) {
+                var newLayerIDs = [event.ID];
+                
+                result.layerIDs.reduce(function (layerIDs, layerID) {
+                    if (!document.layers.byID(layerID) && layerID != event.ID) {
+                        layerIDs.push(layerID);
+                    }
+                    return layerIDs;
+                }, newLayerIDs);
+                
+                return newLayerIDs.reverse().map(function (layerID, index) {
+                    var replacedLayerID = null;
+                    
+                    // For the first new layer, we check whether the previously selected layer is being replaced by
+                    // the new layer.
+                    if (index === 0) {
+                        var lastSelectedLayerID = document.layers.selected.size === 1 ?
+                                document.layers.selected.first().id : null;
+
+                        if (lastSelectedLayerID !== null && result.layerIDs.indexOf(lastSelectedLayerID) === -1) {
+                            replacedLayerID = lastSelectedLayerID;
+                        }
+                    }
+
+                    return {
+                        layerID: layerID,
+                        replacedLayerID: replacedLayerID
+                    };
                 });
-        }
+            })
+            .map(function (result) {
+                var newLayerID = result.layerID,
+                    layer = document.layers.byID(newLayerID);
+
+                // If the new layer is already existed (by expanding a libraries graphic that is an embedded SO),
+                // we reset the layer to update its smart object type and position.
+                if (layer) {
+                    return this.transfer(historyActions.newHistoryStateRogueSafe, document.id)
+                        .bind(this)
+                        .then(function () {
+                            return this.transfer(layerActions.resetLayers, document, layer);
+                        })
+                        .then(function () {
+                            return this.transfer(layerActions.resetIndex, document);
```

Comment:
I don't think we should ever need to call this more than once, but it seems from the control structure that it could actually be called N times for N placed layers, resulting in N^2 layers eventually reset. Should this be factored out and only called once? Or do we know for some reason that this condition can only hold for, say, one layer?